### PR TITLE
Convert ML.Data usages of SubComponent to IComponentFactory.

### PR DESCRIPTION
**Pull request by eerhardt**
With this change, all the usages of SubComponent from our main API assemblies are now removed.

The usages that are left are all under:

1. ComponentCatalog/CmdParser/SubComponent.cs and one line in ComponentCreation.cs
2. ML.PipelineInference
3. ML.ResultProcessor
4. ML.Sweeper
5. ML.Maml
6. EntryPoints CodeGen
7. Tests

Working towards #585

### CODE REVIEW

File: src/Microsoft.ML.Data/Commands/CrossValidationCommand.cs [hunk 1]
```diff
@@ -25,14 +25,14 @@ public sealed class CrossValidationCommand : DataCommand.ImplBase<CrossValidatio
         // REVIEW: We need a way to specify different data sets, not just LabeledExamples.
         public sealed class Arguments : DataCommand.ArgumentsBase
         {
-            [Argument(ArgumentType.Multiple, HelpText = "Trainer to use", ShortName = "tr")]
-            public SubComponent<ITrainer, SignatureTrainer> Trainer = new SubComponent<ITrainer, SignatureTrainer>("AveragedPerceptron");
```
Comment (eerhardt) [inline src/Microsoft.ML.Data/Commands/CrossValidationCommand.cs]: Note: I am taking option (1) here from https://github.com/dotnet/machinelearning/issues/726

Comment (TomFinley) [reply src/Microsoft.ML.Data/Commands/CrossValidationCommand.cs]: Sounds good to me, thanks @eerhardt .

---
In reply to: [212738893](https://github.com/dotnet/machinelearning/pull/729#discussion_r212738893) [](ancestors = 212738893)

File: src/Microsoft.ML.Data/Commands/SaveDataCommand.cs [hunk 2]
```diff
@@ -12,6 +12,7 @@
 using Microsoft.ML.Runtime.Data;
 using Microsoft.ML.Runtime.Data.IO;
 using Microsoft.ML.Runtime.Internal.Utilities;
+using Microsoft.ML.Runtime.EntryPoints;
```
Comment (codemzs) [inline src/Microsoft.ML.Data/Commands/SaveDataCommand.cs]: sort

File: src/Microsoft.ML.Data/Commands/ScoreCommand.cs [hunk 3]
```diff
@@ -135,14 +136,24 @@ private void RunCore(IChannel ch)
...
                 var isText = ext == ".txt" || ext == ".tlc";
-                saver = new SubComponent<IDataSaver, SignatureDataSaver>(isText ? "TextSaver" : "BinarySaver");
+                if (isText)
+                {
+                    writer = new TextSaver(Host, new TextSaver.Arguments());
+                }
+                else
+                {
+                    writer = new BinarySaver(Host, new BinarySaver.Arguments());
+                }
```
[the remark this answers is not in the document]
Comment (eerhardt) [reply src/Microsoft.ML.Data/Commands/ScoreCommand.cs]: Either is "allowed" by our coding guidelines: https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md.

I generally always put braces to ensure the intent is explicit. For example, so bugs like [this](https://www.imperialviolet.org/2014/02/22/applebug.html) don't happen.

File: src/Microsoft.ML.Data/Commands/SaveDataCommand.cs (hunk 2 above)
Comment (eerhardt) [reply src/Microsoft.ML.Data/Commands/SaveDataCommand.cs]: fixed.

**Comment by eerhardt** [comment]
@dotnet-bot test this please

**Pull request merged**